libmscore/scale.cpp

```diff
@@ -0,0 +1,478 @@
+            if (originalNotes[tpc - TPC_MIN].contains('/')) {
+                  QStringList values = originalNotes[tpc - TPC_MIN].split('/');
+                  if (values.size() != 2 || values[1].toDouble() == 0) {
+                        MScore::lastError = "Unsupported Scala file format.";
+                        return false;
+                        }
+                  }
+            if (tpc < minTpc)      minTpc = tpc;      // keep note of tpc range
+            if (tpc > maxTpc)      maxTpc = tpc;
+            }
+      computeTunings();
+
+      return true;
+      }
+
+//---------------------------------------------------------
+//   convertValue
+//---------------------------------------------------------
+
+float Scale::convertValue(QString value, int mode)
+      {
+      float floatValue = 0;
+      if (value.contains('/')) {
+            if (mode == ABSOLUTE_CENTS || mode == ABSOLUTE_FREQUENCY) {
+                  QStringList values = value.split('/');
+                  if (values[1].toDouble() == 0)
+                        values[1] = "1";
+                  floatValue = 1200 * log(values[0].toDouble() / values[1].toDouble()) / log(2);
+                  }
+            }
+      else if (mode == ABSOLUTE_CENTS)
+            floatValue = value.toDouble();
+      else if (mode == DELTA_CENTS)
+            floatValue = value.toDouble();
+      else if (mode == ABSOLUTE_FREQUENCY)
+            floatValue = 1200 * log(value.toDouble()) / log(2);
+      else
+            return 0;
+
+      return floatValue;
+      }
+
+//---------------------------------------------------------
+//   ConvertNoteValue
+//---------------------------------------------------------
+
+float Scale::convertNoteValue(QString noteValue, int tpc, int mode, bool storeFifths, float prevValue)
+      {
+      float value = convertValue(noteValue, mode);
+
+      if (storeFifths)
+            value += prevValue;
+
+      if (mode == ABSOLUTE_CENTS || mode == ABSOLUTE_FREQUENCY) {
+            // convert from offset within an octave to a delta from 12EDO
+            if (!storeFifths)
+                  value -= getStandardNoteValue(tpc);
+            else
+                  value -= getDifference(getStandardNoteValue(tpc),
+                        getStandardNoteValue(prevNote(tpc)));
+            }
+
+      value = (value >= CENTS_PER_OCTAVE) ? value - CENTS_PER_OCTAVE : value;
+      return value;
+      }
+
+//---------------------------------------------------------
+//   getStandardNoteValue
+//---------------------------------------------------------
+
+int Scale::getStandardNoteValue(int tpc)
+      {
+      return ((TPC_FIRST_STEP + (tpc - TPC_MIN) * PITCH_DELTA_FIFTH) * CENTS_PER_SEMITONE) % CENTS_PER_OCTAVE;
+      }
+
+//---------------------------------------------------------
+//   computeTunings
+//---------------------------------------------------------
+
+void Scale::computeTunings(int mode, bool storeFifths)
+      {
```

Comment:
Maybe you should leave this comment on the previous PR for editing the scale?, and here comment only on the saving feature?